#Prime Factoring by QUBO
Now we solve simple prime factoring by using QUBO matrix to solve the minimum problem.

##Factoring 15
Now we try factoring 15 to p and q.
p and q are prime factors

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=15%20%3D%20p*q">

To get p and q we solve the equation below.

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%20(15%20-p%20*%20q)%5E2">

p and q also can write as decimal number using qubits which has binary value.

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=p%20%3D%201%2B2q_0%2B4q_1">

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=q%20%3D%201%2B2q_2">

substitute these and we get equation below.

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%20%5C%7B15-(1%2B2q_0%2B4q_1)(1%2B2q_2)%5C%7D%5E2">

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%2016q_0%5E2q_2%5E2%20%2B%2016q_0%5E2q_2%20%2B%204q_0%5E2%20%2B%2064q_0q_1q_2%5E2%20%2B%2064q_0q_1q_2%20%2B%2016q_0q_1%20%2B%2016q_0q_2%5E2%20-%20104q_0q_2%20-%2056q_0%20%2B%2064q_1%5E2q_2%5E2%20%2B%2064q_1%5E2q_2%20%2B%2016q_1%5E2%20%2B%2032q_1q_2%5E2%20-%20208q_1q_2%20-%20112q_1%20%2B%204q_2%5E2%20-%2056q_2%20%2B%20196">

as the binary numer rules we get.

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%20128q_0q_1q_2%20%2B%2016q_0q_1%20-%2056q_0q_2%20-%2052q_0%20-%2048q_1q_2%20-%2096q_1%20-%2052q_2%20%2B%20196">




##Using boolean reduction to the 3-body interaction

Here we have 3-body interaction which the ising model machines cannot solve. We have to reduce the number of interaction from 3-body to 2-body.

Now we prepare q1*q2 = q3 and put the penarty term at the end of equation.

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%20128q_0q_3%20%2B%2016q_0q_1%20-%2056q_0q_2%20-%2052q_0%20-%2048q_1q_2%20-%2096q_1%20-%2052q_2%20%2B%20196%20%2B%20%5Cdelta(3q_3%2Bq_1q_2-2q_1q_3-2q_2q_3)">

Now we use Gamma = 128 and solve this equation creating QUBO matrix which has coefficiences in the equation as the elements in the matrix.

In [0]:
import blueqat.opt as wq
a = wq.opt()
a.qubo = [[-52,16,-56,128],[0,-96,80,-256],[0,0,-52,-256],[0,0,0,384]]
a.sa()

1.6109580993652344


[0, 1, 1, 1]

In [0]:
print(a.E[-1])

-196.0


Now we get []0,0,0,1] and it shows

p = 1+0+4=5
q = 1+2 = 3

total energy is -196 and this is equal to 0 if you substitute this value to the energy function.

And finally we solved the problem using QUBO and minimum problem to solve simple factoring.